pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -0,0 +1,696 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import "forge-std/Test.sol";
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+import { IAuthentication } from "@balancer-labs/v3-interfaces/contracts/solidity-utils/helpers/IAuthentication.sol";
+import { IProtocolFeeController } from "@balancer-labs/v3-interfaces/contracts/vault/IProtocolFeeController.sol";
+
+import { ArrayHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/ArrayHelpers.sol";
+import { ERC20TestToken } from "@balancer-labs/v3-solidity-utils/contracts/test/ERC20TestToken.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+import { BaseVaultTest } from "./utils/BaseVaultTest.sol";
+
+contract E2eSwapTest is BaseVaultTest {
+    using ArrayHelpers for *;
+    using FixedPoint for uint256;
+
+    ERC20TestToken internal tokenA;
+    ERC20TestToken internal tokenB;
+    uint256 internal tokenAIdx;
+    uint256 internal tokenBIdx;
+
+    uint256 internal decimalsTokenA;
+    uint256 internal decimalsTokenB;
+    uint256 internal poolInitAmountTokenA;
+    uint256 internal poolInitAmountTokenB;
+
+    address internal sender;
+    address internal poolCreator;
+
+    uint256 internal minPoolSwapFeePercentage;
+    uint256 internal maxPoolSwapFeePercentage;
+
+    uint256 internal minSwapAmountTokenA;
+    uint256 internal maxSwapAmountTokenA;
+
+    uint256 internal minSwapAmountTokenB;
+    uint256 internal maxSwapAmountTokenB;
+
+    function setUp() public virtual override {
+        BaseVaultTest.setUp();
+
+        // Tokens must be set before other variables, so the variables can be calculated based on tokens.
+        setUpTokens();
+
+        (tokenAIdx, tokenBIdx) = getSortedIndexes(address(tokenA), address(tokenB));
+
+        decimalsTokenA = IERC20Metadata(address(tokenA)).decimals();
+        decimalsTokenB = IERC20Metadata(address(tokenB)).decimals();
+        poolInitAmountTokenA = poolInitAmount.mulDown(10 ** decimalsTokenA);
+        poolInitAmountTokenB = poolInitAmount.mulDown(10 ** decimalsTokenB);
+
+        setUpVariables();
+        createAndInitCustomPool();
+
+        _donateToVault();
```

Comment:
ah we're setting the balances manually. 

For the purposes of this test I think this is OK; it shouldn't affect the outcome.
Was this just a shortcut, or is it that the invariant ratio for add liquidity unbalanced makes setting the initial conditions harder?

Reply:
It was a shortcut, just a way to change pool balances without initializing the pool again. I just change the pool balances since the vault already have reserves for the token.